### ISSUE TITLE
Problem with MNIST iterator

### ISSUE THREAD

**Issue by Bumblebee1964** [dmlc/MXNet.cpp#77]
I run MXNet with MXNet.cpp on Windows 10. The mlp.cpp example runs ok for both GPU and CPU. But running the mnist / lenet example with dataiter, creating the MNIST dataIter throws an error. 

int main(int argc, char const *argv[]) {
  /*setup basic configs*/
  int W = 28;
  int H = 28;
  int batch_size = 128;
  int max_epoch = 100;
  float learning_rate = 1e-4;
  float weight_decay = 1e-4;

  /*char cwd[1024];
  if (_getcwd(cwd, sizeof(cwd)) != NULL)
	  fprintf(stdout, "Current working dir: %s\n", cwd);
  else
	  perror("getcwd() error");*/

  auto lenet = LenetSymbol();
  std::map<string, NDArray> args_map;

  args_map["data"] = NDArray(Shape(batch_size, 1, W, H), Context::cpu());
  args_map["data_label"] = NDArray(Shape(batch_size), Context::cpu());
  lenet.InferArgsMap(Context::cpu(), &args_map, args_map);

  args_map["fc1_w"] = NDArray(Shape(500, 4 * 4 * 50), Context::cpu());
  NDArray::SampleGaussian(0, 1, &args_map["fc1_w"]);
  args_map["fc2_b"] = NDArray(Shape(10), Context::cpu());
  args_map["fc2_b"] = 0;

  auto train_iter = MXDataIter("MNISTIter")
	       //.SetParam("image", "C:\\Temp\\MXNet-DataSets\\MNIST\\train-images-idx3-ubyte")
	       //.SetParam("label", "C:\\Temp\\MXNet-DataSets\\MNIST\\train-labels-idx1-ubyte")
	        .SetParam("image", "./MNIST/train-images-idx3-ubyte")
	        .SetParam("label", "./MNIST/train-labels-idx1-ubyte")
	  .SetParam("batch_size", batch_size)
      .SetParam("shuffle", 1)
      .SetParam("flat", 0)
      .CreateDataIter();
  auto val_iter = MXDataIter("MNISTIter")
	  //      .SetParam("image", "C:\\Temp\\MXNet-DataSets\\MNIST\\t10k-images-idx3-ubyte")
	  //      .SetParam("label", "C:\\Temp\\MXNet-DataSets\\MNIST\\t10k-labels-idx1-ubyte")
	        .SetParam("image", "./MNIST/t10k-images-idx3-ubyte")
	        .SetParam("label", "./MNIST/t10k-labels-idx1-ubyte")
	  .CreateDataIter();

I do not know how the file path on a Windows should be set but I tried both the \\ and / folder separator and relative as absolute paths. I always see in the output of the debugger the message:
"Exception thrown at 0x00007FFECDE27788 in MxnetTestApp.exe: Microsoft C++ exception: dmlc::Error at memory location 0x000000B0E0AFD9D8"

The error is the same (including the memeory address) for both the training and validation DataIter.

Any help? Is there a log file generated and if so, where can I find it?

**Comment by Bumblebee1964** [comment]
In addition to the problem above, I managed to debug in the MxNet code. In source file local_filesys.cc there is the following method:

void LocalFileSystem::ListDirectory(const URI &path, std::vector<FileInfo> *out_list) {
#ifndef _MSC_VER
  DIR *dir = opendir(path.name.c_str());
  if (dir == NULL) {
    int errsv = errno;
    LOG(FATAL) << "LocalFileSystem.ListDirectory " << path.str()
               <<" error: " << strerror(errsv);
  }
  out_list->clear();
  struct dirent *ent;
  /* print all the files and directories within directory */
  while ((ent = readdir(dir)) != NULL) {
    if (!strcmp(ent->d_name, ".")) continue;
    if (!strcmp(ent->d_name, "..")) continue;
    URI pp = path;
    if (pp.name[pp.name.length() - 1] != '/') {
      pp.name += '/';
    }
    pp.name += ent->d_name;
    out_list->push_back(GetPathInfo(pp));
  }
  closedir(dir);
#else
  WIN32_FIND_DATA fd;
  std::string pattern = path.name + "/*";
  HANDLE handle = FindFirstFile(pattern.c_str(), &fd);
  if (handle == INVALID_HANDLE_VALUE) {
    int errsv = GetLastError();
    LOG(FATAL) << "LocalFileSystem.ListDirectory " << path.str()
               << " error: " << strerror(errsv);
  }
  do {
    if (strcmp(fd.cFileName, ".") && strcmp(fd.cFileName, "..")) {
      URI pp = path;
      char clast = pp.name[pp.name.length() - 1];
      if (pp.name == ".") {
        pp.name = fd.cFileName;
      } else if (clast != '/' && clast != '\\') {
        pp.name += '/';
        pp.name += fd.cFileName;
      }
      out_list->push_back(GetPathInfo(pp));
    }
  }  while (FindNextFile(handle, &fd));
  FindClose(handle);
#endif
}

For Windows systems, I guess the character '/' in the serach pattern is illegal and should be a '\'. I am right?????

**Comment by Bumblebee1964** [comment]
Sorry, the formatter of the textbox is messing up:
For Windows systems, I guess the character **/** in the search pattern is illegal and should be a **backslash**. I am right?????
